File: frontend/src/app/api/portfolio.ts
import { API_BASE_URL, apiFetch, getErrorMessage } from "./client";

export type PortfolioStatus = "작성중" | "보완 필요" | "정리 완료";

export type PortfolioProjectApiItem = {
  id: number;
  title: string;
  publishedPostId: number | null;
  publishedPostIsPublic: boolean | null;
  publishStatus?: "created" | "updated" | "unchanged" | null;
  repoFullName: string;
  githubBranch: string;
  githubUrl: string;
  summary: string | null;
  techStack: string[];
  readmeSummary: string | null;
  readmeContentSaved: boolean;
  recentCommitSummary: string[];
  githubCommits: {
    sha: string;
    message: string;
    authorName: string | null;
    committedAt: string | null;
    htmlUrl: string | null;
  }[];
  savedPortfolioDraft: string | null;
  savedInterviewQuestions: string | null;
  portfolioStatus: PortfolioStatus;
  coachFeedbackStatus: string;
  githubConnected: boolean;
  aiDraftSaved: boolean;
  aiInterviewSaved: boolean;
  lastCommitAt: string | null;
  linkedPostIds: number[];
  linkedRecordCount: number;
  createdAt: string;
  updatedAt: string;
};

export type PortfolioProjectListResponse = {
  items: PortfolioProjectApiItem[];
  total: number;
};

export type PortfolioProjectCreatePayload = {
  githubUrl: string;
  title?: string;
  summary?: string;
  techStack?: string[];
};

export type PortfolioProjectUpdatePayload = {
  title?: string;
  summary?: string;
  techStack?: string[];
  portfolioStatus?: PortfolioStatus;
  savedPortfolioDraft?: string;
  savedInterviewQuestions?: string;
};

type RawPortfolioProjectApiItem = Partial<PortfolioProjectApiItem> & {
  id: number;
  title?: string | null;
};

function normalizeStringArray(value: unknown): string[] {
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : [];
}

function normalizeNumberArray(value: unknown): number[] {
  return Array.isArray(value) ? value.filter((item): item is number => typeof item === "number") : [];
}

function normalizeGithubCommits(value: unknown): PortfolioProjectApiItem["githubCommits"] {
  if (!Array.isArray(value)) {
    return [];
  }

  return value
    .filter((commit): commit is Record<string, unknown> => commit !== null && typeof commit === "object")
    .map((commit) => ({
      sha: typeof commit.sha === "string" ? commit.sha : "",
      message: typeof commit.message === "string" ? commit.message : "",
      authorName: typeof commit.authorName === "string" ? commit.authorName : null,
      committedAt: typeof commit.committedAt === "string" ? commit.committedAt : null,
      htmlUrl: typeof commit.htmlUrl === "string" ? commit.htmlUrl : null,
    }))
    .filter((commit) => commit.sha && commit.message);
}

function normalizePortfolioProject(item: RawPortfolioProjectApiItem): PortfolioProjectApiItem {
  // 백엔드가 재시작되기 전이거나 과거 응답 캐시가 남아 있으면 새 필드가 없을 수 있다.
  // 화면 컴포넌트가 배열/boolean 필드를 바로 읽어도 런타임에서 깨지지 않도록 API 경계에서 기본값을 맞춘다.
  return {
    id: item.id,
    title: item.title ?? "이름 없는 프로젝트",
    publishedPostId: item.publishedPostId ?? null,
    publishedPostIsPublic: item.publishedPostIsPublic ?? null,
    publishStatus: item.publishStatus ?? null,
    repoFullName: item.repoFullName ?? "",
    githubBranch: item.githubBranch ?? "main",
    githubUrl: item.githubUrl ?? "",
    summary: item.summary ?? null,
    techStack: normalizeStringArray(item.techStack),
    readmeSummary: item.readmeSummary ?? null,
    readmeContentSaved: Boolean(item.readmeContentSaved),
    recentCommitSummary: normalizeStringArray(item.recentCommitSummary),
    githubCommits: normalizeGithubCommits(item.githubCommits),
    savedPortfolioDraft: item.savedPortfolioDraft ?? null,
    savedInterviewQuestions: item.savedInterviewQuestions ?? null,
    portfolioStatus: item.portfolioStatus ?? "작성중",
    coachFeedbackStatus: item.coachFeedbackStatus ?? "요청 전",
    githubConnected: Boolean(item.githubConnected),
    aiDraftSaved: Boolean(item.aiDraftSaved),
    aiInterviewSaved: Boolean(item.aiInterviewSaved),
    lastCommitAt: item.lastCommitAt ?? null,
    linkedPostIds: normalizeNumberArray(item.linkedPostIds),
    linkedRecordCount: item.linkedRecordCount ?? normalizeNumberArray(item.linkedPostIds).length,
    createdAt: item.createdAt ?? new Date().toISOString(),
    updatedAt: item.updatedAt ?? new Date().toISOString(),
  };
}

function normalizePortfolioProjectList(data: unknown): PortfolioProjectListResponse {
  const rawItems =
    data !== null &&
    typeof data === "object" &&
    "items" in data &&
    Array.isArray((data as { items: unknown }).items)
      ? (data as { items: RawPortfolioProjectApiItem[] }).items
      : [];

  return {
    items: rawItems.map(normalizePortfolioProject),
    total:
      data !== null && typeof data === "object" && "total" in data && typeof (data as { total: unknown }).total === "number"
        ? (data as { total: number }).total
        : rawItems.length,
  };
}

export async function getPortfolioProjects(): Promise<PortfolioProjectListResponse> {
  const response = await apiFetch(`${API_BASE_URL}/portfolio/projects`, {
    credentials: "include",
  });

  if (!response.ok) {
    const message = await getErrorMessage(response);
    throw new Error(message);
  }

  const data = await response.json();

  return normalizePortfolioProjectList(data);
}

export async function createPortfolioProject(payload: PortfolioProjectCreatePayload): Promise<PortfolioProjectApiItem> {
  const response = await apiFetch(`${API_BASE_URL}/portfolio/projects`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const message = await getErrorMessage(response);
    throw new Error(message);
  }

  const data = await response.json();

  return normalizePortfolioProject(data);
}

export async function updatePortfolioProject(
  projectId: number,
  payload: PortfolioProjectUpdatePayload,
): Promise<PortfolioProjectApiItem> {
  const response = await apiFetch(`${API_BASE_URL}/portfolio/projects/${projectId}`, {
    method: "PATCH",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const message = await getErrorMessage(response);
    throw new Error(message);
  }

  const data = await response.json();

  return normalizePortfolioProject(data);
}

export async function linkPortfolioProjectPosts(projectId: number, postIds: number[]): Promise<PortfolioProjectApiItem> {
  const response = await apiFetch(`${API_BASE_URL}/portfolio/projects/${projectId}/posts`, {
    method: "PUT",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ postIds }),
  });

  if (!response.ok) {
    const message = await getErrorMessage(response);
    throw new Error(message);
  }

  const data = await response.json();

  return normalizePortfolioProject(data);
}

export async function refreshPortfolioProjectGithubInfo(projectId: number): Promise<PortfolioProjectApiItem> {
  const response = await apiFetch(`${API_BASE_URL}/portfolio/projects/${projectId}/github/refresh`, {
    method: "POST",
    credentials: "include",
  });

  if (!response.ok) {
    const message = await getErrorMessage(response);
    throw new Error(message);
  }

  const data = await response.json();

  return normalizePortfolioProject(data);
}

export async function publishPortfolioProjectPost(projectId: number, isPublic: boolean): Promise<PortfolioProjectApiItem> {
  const response = await apiFetch(`${API_BASE_URL}/portfolio/projects/${projectId}/publish-post`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ isPublic }),
  });

  if (!response.ok) {
    const message = await getErrorMessage(response);
    throw new Error(message);
  }

  const data = await response.json();

  return normalizePortfolioProject(data);
}

export async function deletePortfolioProject(projectId: number): Promise<void> {
  const response = await apiFetch(`${API_BASE_URL}/portfolio/projects/${projectId}`, {
    method: "DELETE",
    credentials: "include",
  });

  if (!response.ok) {
    const message = await getErrorMessage(response);
    throw new Error(message);
  }
}

File: backend/app/schemas/portfolio.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class FrontendResponseModel(BaseModel):
    model_config = ConfigDict(populate_by_name=True)


class PortfolioProjectCreateRequest(FrontendResponseModel):
    github_url: str = Field(alias="githubUrl", min_length=1, max_length=500)
    title: str | None = Field(default=None, max_length=200)
    summary: str | None = None
    tech_stack: list[str] = Field(default_factory=list, alias="techStack", max_length=20)


class PortfolioProjectUpdateRequest(FrontendResponseModel):
    title: str | None = Field(default=None, max_length=200)
    summary: str | None = None
    tech_stack: list[str] | None = Field(default=None, alias="techStack", max_length=20)
    portfolio_status: str | None = Field(default=None, alias="portfolioStatus", max_length=30)
    saved_portfolio_draft: str | None = Field(default=None, alias="savedPortfolioDraft")
    saved_interview_questions: str | None = Field(default=None, alias="savedInterviewQuestions")


class PortfolioProjectPostLinkRequest(FrontendResponseModel):
    post_ids: list[int] = Field(default_factory=list, alias="postIds")


class PortfolioProjectPublishRequest(FrontendResponseModel):
    is_public: bool = Field(default=True, alias="isPublic")


class GitHubCommitResponse(FrontendResponseModel):
    sha: str
    message: str
    author_name: str | None = Field(default=None, alias="authorName")
    committed_at: datetime | None = Field(default=None, alias="committedAt")
    html_url: str | None = Field(default=None, alias="htmlUrl")


class PortfolioProjectResponse(FrontendResponseModel):
    id: int
    title: str
    published_post_id: int | None = Field(alias="publishedPostId")
    published_post_is_public: bool | None = Field(default=None, alias="publishedPostIsPublic")
    publish_status: str | None = Field(default=None, alias="publishStatus")
    repo_full_name: str = Field(alias="repoFullName")
    github_branch: str = Field(alias="githubBranch")
    github_url: str = Field(alias="githubUrl")
    summary: str | None
    tech_stack: list[str] = Field(alias="techStack")
    readme_summary: str | None = Field(alias="readmeSummary")
    readme_content_saved: bool = Field(alias="readmeContentSaved")
    recent_commit_summary: list[str] = Field(alias="recentCommitSummary")
    github_commits: list[GitHubCommitResponse] = Field(alias="githubCommits")
    saved_portfolio_draft: str | None = Field(alias="savedPortfolioDraft")
    saved_interview_questions: str | None = Field(alias="savedInterviewQuestions")
    portfolio_status: str = Field(alias="portfolioStatus")
    coach_feedback_status: str = Field(alias="coachFeedbackStatus")
    github_connected: bool = Field(alias="githubConnected")
    ai_draft_saved: bool = Field(alias="aiDraftSaved")
    ai_interview_saved: bool = Field(alias="aiInterviewSaved")
    last_commit_at: datetime | None = Field(alias="lastCommitAt")
    linked_post_ids: list[int] = Field(alias="linkedPostIds")
    linked_record_count: int = Field(alias="linkedRecordCount")
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")


class PortfolioProjectListResponse(FrontendResponseModel):
    items: list[PortfolioProjectResponse]
    total: int
